jenkins/spark-tests.sh

```diff
@@ -98,14 +77,105 @@ jps
+--conf spark.dynamicAllocation.enabled=false \
+--conf spark.driver.extraJavaOptions=-Duser.timezone=UTC \
+--conf spark.executor.extraJavaOptions=-Duser.timezone=UTC \
+--conf spark.sql.session.timeZone=UTC"
+
+export SEQ_CONF="--executor-memory 12G \
+--total-executor-cores 6"
+
+# currently we hardcode the parallelism and configs based on our CI node's hardware specs,
+# we can make it dynamically generated if this script is going to be used in other scenarios in the future
+export PARALLEL_CONF="--executor-memory 4G \
+--total-executor-cores 2 \
+--conf spark.executor.cores=2 \
+--conf spark.task.cpus=1 \
+--conf spark.rapids.sql.concurrentGpuTasks=2 \
+--conf spark.rapids.memory.gpu.allocFraction=0.15 \
+--conf spark.rapids.memory.gpu.minAllocFraction=0 \
+--conf spark.rapids.memory.gpu.maxAllocFraction=0.15"
+
+export CUDF_UDF_TEST_ARGS="--conf spark.rapids.memory.gpu.allocFraction=0.1 \
+--conf spark.rapids.memory.gpu.minAllocFraction=0 \
+--conf spark.rapids.python.memory.gpu.allocFraction=0.1 \
+--conf spark.rapids.python.concurrentPythonWorkers=2 \
+--conf spark.executorEnv.PYTHONPATH=${RAPIDS_PLUGIN_JAR} \
+--conf spark.pyspark.python=/opt/conda/bin/python \
+--py-files ${RAPIDS_PLUGIN_JAR}"
+
+export TEST_PARALLEL=0  # disable spark local parallel in run_pyspark_from_build.sh
+export TEST_TYPE="nightly"
+export LOCAL_JAR_PATH=$ARTF_ROOT
+export SCRIPT_PATH="$(pwd -P)"
+export TARGET_DIR="$SCRIPT_PATH/target"
+mkdir -p $TARGET_DIR
+
+run_test() {
+    local TEST=${1//\.py/}
+    local LOG_FILE
+    case $TEST in
+      all)
+        SPARK_SUBMIT_FLAGS="$BASE_SPARK_SUBMIT_ARGS $SEQ_CONF" \
+          ./run_pyspark_from_build.sh
+        ;;
+
+      cudf_udf_test)
+        SPARK_SUBMIT_FLAGS="$BASE_SPARK_SUBMIT_ARGS $SEQ_CONF $CUDF_UDF_TEST_ARGS" \
+          ./run_pyspark_from_build.sh -m cudf_udf --cudf_udf
+        ;;
+
+      cache_serializer)
+        SPARK_SUBMIT_FLAGS="$BASE_SPARK_SUBMIT_ARGS $SEQ_CONF \
+        --conf spark.sql.cache.serializer=com.nvidia.spark.rapids.shims.spark311.ParquetCachedBatchSerializer" \
+          ./run_pyspark_from_build.sh -k cache_test
+        ;;
+
+      *)
+        echo -e "\n\n>>>>> $TEST...\n"
+        LOG_FILE="$TARGET_DIR/$TEST.log"
+        # set dedicated RUN_DIRs here to avoid conflict between parallel tests
+        RUN_DIR="$TARGET_DIR/run_dir_$TEST" \
+          SPARK_SUBMIT_FLAGS="$BASE_SPARK_SUBMIT_ARGS $PARALLEL_CONF" \
+          ./run_pyspark_from_build.sh -k $TEST >"$LOG_FILE" 2>&1
+
+        CODE="$?"
+        if [[ $CODE == "0" ]]; then
+          sed -n -e '/test session starts/,/deselected,/ p' "$LOG_FILE" || true
+        else
+          cat "$LOG_FILE" || true
+        fi
+        return $CODE
+        ;;
+    esac
+}
+export -f run_test
+
+# integration tests
+if [[ $PARALLEL_TEST == "true" ]] && [ -x "$(command -v parallel)" ]; then
+  # put most time-consuming tests at the head of queue
+  time_consuming_tests="join_test.py generate_expr_test.py"
```

Comment:
I thought we were tagging the source to indicate they were slow as in #3241.  Curious why this isn't leveraging that, e.g.: `time_consuming_tests=$(grep -rl pytest.mark.slow_test "$SCRIPT_PATH"/src/main/python

Reply:
`slow_tag` there was for pre-merge parallel split (either time-consuming or high memory usage), mostly to balance test time of two parallel test stages (the tag naming looks ambiguous)
In nightly cases, I want pick time-consuming ones only, and per my tests join_test (3000s) + generate_expr_test (1800s) were the only 2 tests that spent consistently over 15 mins, so I manually put them here to avoid that they were in the middle or tail of task queue

Reply:
Should we split the test cases in `join_test` and `generate_expr_test` so they spread more evenly even without using the `parallel` hack?

Reply:
Yes, split in test-cases level that would be better. I have tested w/ a few non-serious cases-level split but not getting much benefits from it, and this could make it harder for developer to manage the test scenarios. I would like to get back to this if current setup does not fulfill our efficiency requirement